CoffeeChat/Utilities/Time+Helper.swift

```diff
@@ -35,16 +59,119 @@ class Time {
     */
     static func isTommorow(_ date: Date) -> Bool {
+        let calendar = Calendar.current
         if
-            let today = Calendar.current.date(bySettingHour: 0, minute: 0, second: 0, of: Date()),
-            let meetingDate = Calendar.current.date(bySettingHour: 0, minute: 0, second: 0, of: date)
+            let today = calendar.date(bySettingHour: 0, minute: 0, second: 0, of: Date()),
+            let meetingDate = calendar.date(bySettingHour: 0, minute: 0, second: 0, of: date)
         {
-            if let dayDifference = Calendar.current.dateComponents([.day], from: today, to: meetingDate).day {
+            if let dayDifference = calendar.dateComponents([.day], from: today, to: meetingDate).day {
                 return dayDifference <= 1
             }
         }
 
         return false
     }
 
+    /// Returns the `Date` corresponding to next `weekday` at the time specified by `time`
+    static func next(_ weekday: Weekday, at time: Float) -> Date {
+        getWeekday(searchDirection: .forward, weekday: weekday, time: time)
+    }
+
+}
+
+// MARK: Helper Functinos
+extension Time {
+
+    /**
+    Converts a float time to 2 integers representing hours and minutes.
+    Only returns either on the hour times or half hours
+
+    # Example:
+    time = 0 -> 12:00 AM
+    time = 1 -> 1:00 AM
+    time = 13.5 -> 1:30 PM
+    time = 23.5 -> 11:30 PM
+    */
+    private static func floatTimeToHoursMinutes(time: Float) -> (hours: Int, minutes: Int) {
+        let hours = Int(time)
+        let minutes = time.rounded() > time ? 30 : 0
+        return (hours, minutes)
+    }
+
+    private static func getWeekDaysInEnglish() -> [String] {
+        var calendar = Calendar(identifier: .gregorian)
+        calendar.locale = Locale(identifier: "en_US_POSIX")
+        return calendar.weekdaySymbols
+    }
+
+    /**
+    Returns the next or previous `weekday` with a time based on `time`
+    The search for the weekday includes today.
+    # Example:
+    If today is Monday 3:00 PM, `getWeekday(.forward, .sunday, 2)` returns next sunday (6 days from now) at 2:00 AM
+    If today is Sunday 5:00 PM, `getWeekday(.forward, .sunday, 18)` returns **today** at 6:00 PM
+
+    If today is Tuesday 3:00 PM, `getWeekday(.backward, .sunday, 2)` returns the previous sunday (2 days ago) at 2:00 AM
+    If today is Sunday 5:00 PM, `getWeekday(.backward, .sunday, 13)` returns **today** at 1:00 PM
+    */
+    private static func getWeekday(searchDirection: Calendar.SearchDirection, weekday: Weekday, time: Float) -> Date {
+        let calendar = Calendar.current
+
+        // Get the index of the `weekday`
+        let dayName = weekday.rawValue
+        let weekdaysName = Time.getWeekDaysInEnglish().map { $0.lowercased() }
+        guard var searchWeekdayIndex = weekdaysName.firstIndex(of: dayName) else {
+            fatalError("weekday symbol should be in the form \(weekdaysName)")
```

Comment:
This is good for debugging, but not for production. We don't want the app crashing on users. Can this be handled gracefully?

Reply:
yea, i think we should try our best to stay away from raising fatal errors (even though this likely won't happen)